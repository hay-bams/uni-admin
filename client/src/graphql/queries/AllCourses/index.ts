import { gql } from '@apollo/client';

export const All_COURSES = gql`
  query AllCourses {
    allCourses {
      id
      name
      totalSeats
      status
    }
  }
`;
