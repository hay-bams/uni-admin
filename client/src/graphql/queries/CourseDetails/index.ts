import { gql } from '@apollo/client';

export const COURSE_DETAILS = gql`
  query courseDetails($courseId: ID!) {
    courseDetails(courseId: $courseId) {
      id
      name
      status
      totalSeats
    }
  }
`;
