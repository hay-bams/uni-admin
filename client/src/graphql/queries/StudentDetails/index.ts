import { gql } from "@apollo/client";

export const STUDENT_DETAILS = gql`
  query studentDetails ($id: ID!) {
    studentDetails(id: $id) {
      id
      name
      courses {
        id
        name
        totalSeats
        status
      }
    }
  }
`;
