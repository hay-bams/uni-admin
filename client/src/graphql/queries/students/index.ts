import { gql } from '@apollo/client';

export const STUDENT = gql`
  query Students($all: String, $page: Int, $limit: Int) {
    students(all: $all, page: $page, limit: $limit) {
      total
      results {
        id
        studentID
        dob
        status
        name
        email
        country
      }
    }
  }
`;
