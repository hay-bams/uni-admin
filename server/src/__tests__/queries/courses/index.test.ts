/* eslint-disable @typescript-eslint/no-non-null-assertion */
import jwt from 'jsonwebtoken';
import { COURSES_DATA, USER_DATA } from '../../test_data';
import { createTestServer } from '../../helpers';
import { TEST_All_COURSES_QUERY, TEST_SINGLE_COURSE_QUERY } from '../queries';

describe('Courses Queries', () => {
  test('should return all courses', async () => {
    const { query } = createTestServer({
      db: {
        courses: {
          find: jest.fn(() => ({
            skip: jest.fn(() => ({
              limit: jest.fn(() => ({
                count: jest.fn(() => 4),
                toArray: jest.fn(() => COURSES_DATA),
              })),
            })),
          })),
        },
        users: {
          findOne: jest.fn(() => USER_DATA[0]),
        },
      },
      req: {
        get: jest.fn(() => 'skskkdkdk'),
        signedCookies: {
          admin: USER_DATA[0]._id,
        },
      },
      res: {
        clearCookie: jest.fn(),
      },
    });

    jest.spyOn(jwt, 'verify').mockImplementation(
      jest.fn(() => ({
        admin: USER_DATA[0]._id,
      }))
    );

    const res = await query({ query: TEST_All_COURSES_QUERY });

    expect(res.data.allCourses.results.length).toBe(COURSES_DATA.length);
    expect(res.data.allCourses.results[0]).toHaveProperty('id');
    expect(res).toMatchSnapshot();
  });

  test('should return a single courses', async () => {
    const { query } = createTestServer({
      db: {
        courses: {
          findOne: jest.fn(() => COURSES_DATA[0]),
        },
        users: {
          findOne: jest.fn(() => USER_DATA[0]),
        },
      },
      req: {
        get: jest.fn(() => 'skskkdkdk'),
        signedCookies: {
          admin: USER_DATA[0]._id,
        },
      },
      res: {
        clearCookie: jest.fn(),
      },
    });

    jest.spyOn(jwt, 'verify').mockImplementation(
      jest.fn(() => ({
        admin: USER_DATA[0]._id,
      }))
    );

    const res = await query({
      query: TEST_SINGLE_COURSE_QUERY,
      variables: { courseId: COURSES_DATA[0]._id.toString() },
    });

    expect(res.data.courseDetails).toHaveProperty('id');
    expect(res.data.courseDetails).toHaveProperty('name');
    expect(res).toMatchSnapshot();
  });

  test('should throw an error if a single course is not found', async () => {
    const { query } = createTestServer({
      db: {
        courses: {
          findOne: jest.fn(() => null),
        },
        users: {
          findOne: jest.fn(() => USER_DATA[0]),
        },
      },
      req: {
        get: jest.fn(() => 'skskkdkdk'),
        signedCookies: {
          admin: USER_DATA[0]._id,
        },
      },
      res: {
        clearCookie: jest.fn(),
      },
    });

    jest.spyOn(jwt, 'verify').mockImplementation(
      jest.fn(() => ({
        admin: USER_DATA[0]._id,
      }))
    );

    const res = await query({
      query: TEST_SINGLE_COURSE_QUERY,
      variables: { courseId: COURSES_DATA[0]._id.toString() },
    });

    expect(res.errors?.length).toBeGreaterThan(0);
    expect(res.errors![0]).toHaveProperty('message');
    expect(res).toMatchSnapshot();
  });
});
