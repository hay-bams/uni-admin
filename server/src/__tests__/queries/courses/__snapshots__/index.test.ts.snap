// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Courses Queries should return a single courses 1`] = `
Object {
  "data": Object {
    "courseDetails": Object {
      "id": "5fdab398791e134a5466566f",
      "name": "Aerospace Engineerng",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Courses Queries should return all courses 1`] = `
Object {
  "data": Object {
    "allCourses": Object {
      "results": Array [
        Object {
          "id": "5fdab398791e134a5466566f",
        },
        Object {
          "id": "5fdab4b5eed1fb397feaca14",
        },
        Object {
          "id": "5fdab4bd78cc9321b31faf35",
        },
        Object {
          "id": "5fdab4c4fe1aa062abc46096",
        },
        Object {
          "id": "5fdab4ccba6dcf525408412a",
        },
      ],
      "total": 4,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Courses Queries should throw an error if a single course is not found 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Something went wrong: Error: Course not found],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
